File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def post_order(self, root, moves):
        if root == None:
            return 0
        l = self.post_order(root.left, moves)
        r = self.post_order(root.right, moves)
        moves[0] += abs(l) + abs(r)
        return root.val + r + l - 1

    def distributeCoins(self, root):
        moves = [0]*1
        self.post_order(root, moves)
        return moves[0]
        

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int postOrder(TreeNode* root, int &moves) {
        if(!root) return 0;
        int l = postOrder(root -> left, moves);
        int r = postOrder(root -> right, moves);
        moves += abs(l) + abs(r);
        return root -> val + l + r - 1;
    }
    int distributeCoins(TreeNode* root) {
        int moves = 0;
        postOrder(root, moves);
        return moves;
    }
};